### ISSUE TITLE
Automatically include reference metadata and sequence for TreeTime root

### ISSUE THREAD

**Issue by huddlej** [nextstrain/ncov#700]
**Context**
A common use case for ncov builds is to search for specific sequences in GISAID (e.g., for a specific emerging lineage, state, country, etc.) and create a build for these data. The simplest possible `builds.yaml` for this analysis should be:

```yaml
inputs:
  - name: gisaid-data
     sequences: data/downloaded_sequences.fasta
     metadata: data/downloaded_metadata.tsv
```

In practice, users need to also include metadata and sequences for the reference `Wuhan/Hu-1/2019` because the workflow uses this strain to root the time tree. When users do not manually include this reference, the workflow runs all the way through to the `refine` step before throwing a confusing error about "unsupported rooting mechanisms or root not found".

**Description**
We should include the reference sequence and metadata as inputs to all analysis even if users haven't requested them.

**Possible solution**
Instead of including the reference and metadata early in the workflow, we should force-include it either just before creating subsampled data or just after.

Alternately, we could define a different rooting mechanism that doesn't rely on a specific reference strain.

**Additional context**
Thank you to @BryanTegomoh for reporting this issue!

**Comment by emontag47** [comment]
Hi Nextstrain Team,

I was wondering if I could have this issue assigned to me. My approach would be similar to the possible solution. I would include the reference and metadata before creating the subsampled data to to the time tree. I would also love any feedback on this approach. Thank you!

**Comment by tsibley** [comment]
@emontag47 Assigned! Thanks for volunteering and please feel free to discuss/comment on things here.

**Comment by emontag47** [comment]
Hi Nextstrain Team and @tsibley, 

I am attempting to push some changes to a branch I created to eventually submit a pull request. When I try to push, I get the error 
"remote: Permission to nextstrain/ncov.git denied to emontag47.
fatal: unable to access 'https://github.com/nextstrain/ncov.git/': The requested URL returned error: 403".
After searching for a bit on Google, I found that this error could be due to not having permission to push to the repository. Is there anyway someone could give me permission to push to branches or offer some insight into how I could solve this issue? Thank you!

**Comment by jameshadfield** [comment]
Hi Eric - the common way to do this is to [fork the repo and submit a PR from your fork](https://jarv.is/notes/how-to-pull-request-fork-github/). Thanks in advance for your contribution!

**Comment by emontag47** [comment]
Hi @jameshadfield! I will try that and post here again if I am still having issues. Thank you for your help!

**Comment by huddlej** [comment]
@jsan4christ reported this issue recently to @rneher and myself. San is one of [many people who have experienced this issue](https://discussion.nextstrain.org/search?q=unknown%20root).

Although there is a workaround of including the reference data as a separate input for a given build, this solution is not clearly documented. Ideally, the workflow would be more aware of the presence of the reference as early as possible and handle missing reference data more gracefully by either:

1. telling the user how to modify their build configuration to include the reference
2. automatically adding the references, as needed, and then removing them when they are no longer needed.

Both solutions require the workflow to check for the presence of the root in [the subsampled data](https://github.com/nextstrain/ncov/blob/2de8fbcdbb2991619f9835c5206f02d49cc77e33/workflow/snakemake_rules/main_workflow.smk#L429-L430) or perhaps the original input data.

Both solutions could be implemented as a rule that runs a script that checks for the root in the subsampled data. To implement the first solution, that script would print a helpful error message about how to modify the build config and exit the workflow with an error. If the reference data are present, the script would create a placeholder file that allows the workflow to continue without an error. A major disadvantage to this solution is that it alerts the user about the missing references after the most I/O-intensive parts of the workflow have completed and it requires users to modify their build config and re-run all of those parts again, just to get a single sequence added to their analysis. The rule to check for the reference data could be placed earlier in the workflow where it could check all inputs, but this would be complicated by the many different entry points to the workflow and the potentially large input files that would need to be searched.

To implement the second solution, a similar script would also accept reference sequences and metadata as an input and produce a new set of subsampled data with the reference data added, when the reference data are missing. It would never throw an error and would always produce output FASTA and metadata to be used by the rest of the workflow. This solution does have a disadvantage of modifying the user's data without their express approval.

Augur's align command uses something like the second solution [to ensure that the requested reference is included in an alignment](https://github.com/nextstrain/augur/blob/1b56153541bec214b309f8ec8c7938fdd9160ffe/augur/align.py#L67-L81) and then [provides an option to remove the reference after alignment](https://github.com/nextstrain/augur/blob/1b56153541bec214b309f8ec8c7938fdd9160ffe/augur/align.py#L26). A third solution would be to modify `augur refine` to behave similarly to `augur align`, by accepting a reference sequence and metadata as the root that could be automatically removed after running TreeTime in the case where it wasn't present already.

**Comment by trvrb** [comment]
I like the 1+2 proposal above. Keep in mind that even if we could reliably root on "best" (essentially maximizing root-to-tip correlation) we'd be left with an issue of calling mutations. By including Wuhan root mutations are reliably called relative to this reference. This is important for clade assignment as well as general understanding.

**Comment by emmahodcroft** [comment]
I agree with Trevor - I think we'd invite more confusion if mutation calling started getting out-of-whack with what's standard.

**Comment by sidneymbell** [comment]
This sounds great, thank you @huddlej . 

I would perhaps also consider adding a conditional to the proximity calculation script that excludes the reference (i.e., don't enrich for samples similar to the ref sequence).

**Comment by huddlej** [comment]
@sidneymbell I was wondering about that point in @danrlu's Slack thread. I would expect the reference to get ignored during proximity calculations based on [the `--ignore-seqs` that specifies the root strain name](https://github.com/nextstrain/ncov/blob/58ca2bbd166df6d9745939ecfe2e294592de3688/workflow/snakemake_rules/main_workflow.smk#L372). Do you find that you still see the root strain (`Wuhan/Hu-1/2019`) in the proximity scores or perhaps you see the other Wuhan reference strain that we use for GenBank-based builds (`Wuhan/WH01/2019`)?

**Comment by danrlu** [comment]
I'm learning so much through this discussion 😆 I didn't know `--ignore-seqs Wuhan/Hu-1/2019` is in place by default (I see it in our run log). We removed all ref strain from `include.txt` just so we don't get sequences pulled due to proximity to the root, but in fact we only needed to remove the other 2 ref sequences (such as `Wuhan/WH01/2019`), and can keep `Wuhan/Hu-1/2019` in `include.txt` 🤩 But currently we also repurposed our free-of-reference `include.txt` for record keeping and metadata overlay template generation. Just want to add our train of thought for reference. Thanks very much for all the improvements!!!
